### PR TITLE
Synonym inference bug

### PR DESCRIPTION
**Pull request by jy14898**
**Description of the change**

Fixes https://github.com/purescript/purescript/issues/4183

---

**Checklist:**

- [x] Added a file to CHANGELOG.d for this PR (see CHANGELOG.d/README.md)
- [x] Added myself to CONTRIBUTORS.md (if this is my first contribution)
- [x] Linked any existing issues or proposals that this pull request should close
- [x] Updated or added relevant documentation
- [x] Added a test for the contribution (if applicable)

### CODE REVIEW

**Review by garyb — APPROVED** [review]
Looks sensible to me :+1:

**Review by JordanMartinez — APPROVED** [review]
So long as CI builds, I approve.

File: src/Language/PureScript/TypeChecker/Kinds.hs [hunk 1]
```diff
@@ -665,8 +665,10 @@ inferTypeSynonym moduleName (ann, tyName, tyArgs, tyBody) = do
     tyArgs' <- for tyArgs . traverse . maybe (freshKind (fst ann)) $ replaceAllTypeSynonyms <=< apply <=< flip checkKind E.kindType
     unifyKinds tyKind' $ foldr ((E.-:>) . snd) kindRes tyArgs'
     bindLocalTypeVariables moduleName (first ProperName <$> tyArgs') $ do
-      tyBodyAndKind <- inferKind tyBody
+      tyBodyAndKind <- secondM apply =<< inferKind tyBody
       instantiateKind tyBodyAndKind =<< apply kindRes
+  where
+    secondM f (a,b) = (a,) <$> f b
```
Comment (kritzcreek) [inline src/Language/PureScript/TypeChecker/Kinds.hs:671]: Isn't this just `traverse`?

**Comment by rhendric** [comment]
This specific test case would also be fixed by using `apply` on the `TUnknown` created on line 260, in `inferAppKind`, instead of changing `inferTypeSynonym`. These solutions have different scopes of impact; there are some code paths that go through `inferAppKind` and not `inferTypeSynonym`, and vice versa. I don't know if one or the other of these solutions is more correct (or maybe they both are needed?). Does someone else?

**Comment by natefaubion** [comment]
> This specific test case would also be fixed by using apply on the TUnknown created on line 260, in inferAppKind, instead of changing inferTypeSynonym. These solutions have different scopes of impact; there are some code paths that go through inferAppKind and not inferTypeSynonym, and vice versa. I don't know if one or the other of these solutions is more correct (or maybe they both are needed?). Does someone else?

Calling `apply` is idempotent, so you can put it "wherever." The point is just that whenever you expect to make progress with solving, you should always be checking against the latest solutions for the type, since you don't want to unify against outdated information. The only reason I wrote it this way was because it's a pretty literal translation of the paper. I wouldn't do it this way if I were doing it again, since it's clearly error prone. In reality you only need to do this when:
* Solving for an unknown (make sure the unknown isn't already solved, otherwise continue unifying with it's solution).
* Putting a solution into the environment (since you already need to do an occurs check).
* At the point of generalization (leftover unknown are replaced with quantifiers, and you don't want to generalize something that was concretely solved).

I think it's probably a good thing to do this in inferTypeSynonym, since I'd expect you'll want to make sure you're giving the latest to `instantiateKind`.

File: src/Language/PureScript/TypeChecker/Kinds.hs (hunk 1 above)
Comment (jy14898) [reply src/Language/PureScript/TypeChecker/Kinds.hs]: Indeed, hoogle failed me. It's always `traverse`

**Comment by JordanMartinez** [comment]
Shall we merge this PR then? The `traverse` part has been fixed.

**Review by thomashoneyman — APPROVED** [review]

**Pull request merged**